proxy_http.go

```diff
@@ -406,3 +406,50 @@ func defaultFilter(ctx filters.Context, req *http.Request, next filters.Next) (*
 func defaultOnError(ctx filters.Context, req *http.Request, read bool, err error) *http.Response {
 	return nil
 }
+
+// synchronizedDownstream ensures that our read loop and our http.Transport
+// don't encounter race conditions in reading from the buffered downstream
+// connection.
+type synchronizedDownstream struct {
+	*bufio.Reader
+	mx sync.Mutex
+}
+
+func (sd *synchronizedDownstream) Read(b []byte) (int, error) {
+	sd.mx.Lock()
+	n, err := sd.Reader.Read(b)
+	sd.mx.Unlock()
+	return n, err
+}
+
+func (sd *synchronizedDownstream) PeekBuffered() (b []byte) {
+	sd.mx.Lock()
+	buffered := sd.Buffered()
+	if buffered > 0 {
+		b, _ = sd.Peek(buffered)
```

Comment:
Why do we ignore the error here @oxtoacart?

Reply:
This is not new, but we ignore the error because we're only reading the amount that we already know is buffered due to the call on line 427, so Peek can't fail.

Reply:
gotcha thanks